sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/SpannerIO.java

```diff
@@ -102,56 +163,377 @@
+
+    /**
+     * Returns a new {@link SpannerIO.Read} that will write to the specified Cloud Spanner
+     * database.
+     *
+     * <p>Does not modify this object.
+     */
+    public Read withDatabaseId(ValueProvider<String> databaseId) {
+      Builder builder = toBuilder();
+      builder.spannerConfigBuilder().setDatabaseId(databaseId);
+      return builder.build();
+    }
+
+    @VisibleForTesting
+    Read withServiceFactory(ServiceFactory<Spanner, SpannerOptions> serviceFactory) {
+      Builder builder = toBuilder();
+      builder.spannerConfigBuilder().setServiceFactory(serviceFactory);
+      return builder.build();
+    }
+
+    /**
+     * Returns a new {@link SpannerIO.Read} that will read from the specified Cloud Spanner
+     * config.
+     *
+     * <p>Does not modify this object.
+     */
+    public Read withTransaction(PCollectionView<Transaction> transaction) {
+      return toBuilder().setTransaction(transaction).build();
+    }
+
+    @Override
+    public void validate(PipelineOptions options) {
+      getSpannerConfig().validate(options);
+      checkNotNull(
+          getTimestampBound(),
+          "SpannerIO.read() runs in a read only transaction and requires timestamp to be set "
+              + "with withTimestampBound or withTimestamp method");
+
+      if (getQuery() != null) {
+        // TODO: validate query?
+      } else if (getTable() != null) {
+        // Assume read
+        checkNotNull(
+            getColumns(),
+            "For a read operation SpannerIO.read() requires a list of "
+                + "columns to set with withColumns method");
+        checkArgument(
+            !getColumns().isEmpty(),
+            "For a read operation SpannerIO.read() requires a"
+                + " list of columns to set with withColumns method");
+      } else {
+        throw new IllegalArgumentException(
+            "SpannerIO.read() requires configuring query or read operation.");
+      }
+    }
+
+    @Override
+    public PCollection<Struct> expand(PBegin input) {
+      Read config = this;
+      List<PCollectionView<Transaction>> sideInputs = Collections.emptyList();
+      if (getTimestampBound() != null) {
+        PCollectionView<Transaction> transaction =
+            input.apply(createTransaction().withSpannerConfig(getSpannerConfig()));
+        config = config.withTransaction(transaction);
+        sideInputs = Collections.singletonList(transaction);
+      }
+      return input
+          .apply(Create.of(1))
+          .apply(
+              "Execute query", ParDo.of(new NaiveSpannerReadFn(config)).withSideInputs(sideInputs));
+    }
+  }
+
+  /**
+   * A {@link PTransform} that create a transaction.
+   *
+   * @see SpannerIO
+   */
+  @Experimental(Experimental.Kind.SOURCE_SINK)
+  @AutoValue
+  public abstract static class CreateTransaction
+      extends PTransform<PBegin, PCollectionView<Transaction>> {
+
+    private static final long serialVersionUID = 9201734106453817417L;
```

Comment:
serialVersionUID's on PTransforms are also unnecessary. We don't serialize PTransform's using Serializable.